<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Hae Kyung Im" />

<meta name="date" content="2020-06-03" />

<title>QGT-Columbia-analysis-plan</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">QGT-Columbia-HKI</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hakyimlab/QGT-Columbia-HKI">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">QGT-Columbia-analysis-plan</h1>
<h4 class="author">Hae Kyung Im</h4>
<h4 class="date">2020-06-03</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-06-05
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>QGT-Columbia-HKI/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200603code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20200603)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200603code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200603)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomhakyimlabQGTColumbiaHKItree0deee3c4ef98fdd2ee955d953a93ef4a8f1e25a3targetblank0deee3ca"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/hakyimlab/QGT-Columbia-HKI/tree/0deee3c4ef98fdd2ee955d953a93ef4a8f1e25a3" target="_blank">0deee3c</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomhakyimlabQGTColumbiaHKItree0deee3c4ef98fdd2ee955d953a93ef4a8f1e25a3targetblank0deee3ca" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

Unstaged changes:
    Modified:   analysis/analysis_plan.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/0deee3c4ef98fdd2ee955d953a93ef4a8f1e25a3/analysis/analysis_plan.Rmd" target="_blank">0deee3c</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
added TWMR
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI/0deee3c4ef98fdd2ee955d953a93ef4a8f1e25a3/docs/analysis_plan.html" target="_blank">0deee3c</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
added TWMR
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/064f6eeb8a042a0f96f0198269cd6dde92a1a1da/analysis/analysis_plan.Rmd" target="_blank">064f6ee</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
added figures and slides under extras
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI/064f6eeb8a042a0f96f0198269cd6dde92a1a1da/docs/analysis_plan.html" target="_blank">064f6ee</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
added figures and slides under extras
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/1b70eb7f2c25e0ab2e2e0795189fbc829b815d69/analysis/analysis_plan.Rmd" target="_blank">1b70eb7</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
updated prerequisites
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI/1b70eb7f2c25e0ab2e2e0795189fbc829b815d69/docs/analysis_plan.html" target="_blank">1b70eb7</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
updated prerequisites
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/339f40ce0446812e8b6063570927916c29a9cb50/analysis/analysis_plan.Rmd" target="_blank">339f40c</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
added plan
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI/339f40ce0446812e8b6063570927916c29a9cb50/docs/analysis_plan.html" target="_blank">339f40c</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
added plan
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/a3aa03eabde07481472c1e7e9333298978627411/analysis/analysis_plan.Rmd" target="_blank">a3aa03e</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
prerequisites added
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/09f5daef68b4080823d8a3f054e4a8fc29f4a894/analysis/analysis_plan.Rmd" target="_blank">09f5dae</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
fastenloc
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/ccb1167f935aefc2aaeea4e93fe1c1ea284c06d1/analysis/analysis_plan.Rmd" target="_blank">ccb1167</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
knit
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI/ccb1167f935aefc2aaeea4e93fe1c1ea284c06d1/docs/analysis_plan.html" target="_blank">ccb1167</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
knit
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/d427aee02c72daa61f997704fe6db760b4c96519/analysis/analysis_plan.Rmd" target="_blank">d427aee</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
minor comment sort1
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/4ca65b9cb29fb00da7bb35996999f7f73734e0d5/analysis/analysis_plan.Rmd" target="_blank">4ca65b9</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-05
</td>
<td>
edits 2
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/f59cd026fb14b7aa0811300ec41043b93c5ccd1b/analysis/analysis_plan.Rmd" target="_blank">f59cd02</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-04
</td>
<td>
edits
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI/f59cd026fb14b7aa0811300ec41043b93c5ccd1b/docs/analysis_plan.html" target="_blank">f59cd02</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-04
</td>
<td>
edits
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/d65c5558a2c3d732af0ed398d1bc99908b68f331/analysis/analysis_plan.Rmd" target="_blank">d65c555</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-04
</td>
<td>
twmr
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI/d65c5558a2c3d732af0ed398d1bc99908b68f331/docs/analysis_plan.html" target="_blank">d65c555</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-04
</td>
<td>
twmr
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI/682b6e2ba0b24fef25b003b0b289f587ed746d47/docs/analysis_plan.html" target="_blank">682b6e2</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/d3502e82abb1d356884a07f11856dddfd8bb9851/analysis/analysis_plan.Rmd" target="_blank">d3502e8</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-04
</td>
<td>
wflow_publish(“analysis/analysis_plan.Rmd”)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/4420fc7d9c5352115276af310f718e9cc83bd788/analysis/analysis_plan.Rmd" target="_blank">4420fc7</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-04
</td>
<td>
wflow_rename(“analysis/predixcan_analysis.Rmd”, “analysis/analysis_plan.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hakyimlab/QGT-Columbia-HKI/4420fc7d9c5352115276af310f718e9cc83bd788/docs/analysis_plan.html" target="_blank">4420fc7</a>
</td>
<td>
Hae Kyung Im
</td>
<td>
2020-06-04
</td>
<td>
wflow_rename(“analysis/predixcan_analysis.Rmd”, “analysis/analysis_plan.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="set-up" class="section level2">
<h2>set up</h2>
<p>This information is also on the slides - download from Box data and software</p>
<ul>
<li>install anaconda/miniconda</li>
<li>define imlabtools conda environment</li>
<li>download the folder from <a href="">Box here</a> -[ ] TODO: add box link with updated folder this will have copies of all the software repositories and the models
<ul>
<li>download metaxcan repo</li>
<li>download torus repo</li>
<li>download fastenloc repo</li>
<li>download TMWR repo</li>
<li>download prediction models from predictdb.org</li>
</ul></li>
</ul>
<p>workaround: in the experimental Rstudio server (<a href="http://104.154.186.85" class="uri">http://104.154.186.85</a>) For the terminal to work properly we need to run &gt; source /home/rstudio/.bashrc &gt; conda activate imlabtools</p>
</div>
<div id="analysis-plan" class="section level2">
<h2>analysis plan</h2>
<ul>
<li>predict whole blood expression</li>
<li>run association between predicted expression and a simulated phenotype</li>
<li>calculate association between expression levels and coronary artery disease risk using s-predixcan</li>
<li>fine-map the coronary artery disease gwas results using torus (need some preformatting)</li>
<li>calculate colocalization probability using fastenloc</li>
<li>run transcriptome-wide mendelian randomization in one locus of interest</li>
</ul>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>── Attaching packages ───────────────────────────────────────────────────────────────────────── tidyverse 1.3.0 ──</code></pre>
<pre><code>✓ ggplot2 3.2.1     ✓ purrr   0.3.3
✓ tibble  2.1.3     ✓ dplyr   0.8.3
✓ tidyr   1.0.2     ✓ stringr 1.4.0
✓ readr   1.3.1     ✓ forcats 0.4.0</code></pre>
<pre><code>── Conflicts ──────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()</code></pre>
</div>
<div id="preliminary-information" class="section level1">
<h1>Preliminary information</h1>
<p>Data and copies of repositories can be downloaded from <a href="https://uchicago.box.com/s/zhapf2zfxcpj7thvq4sjnqale3emleum">Box here</a></p>
<p>The latest version of the analysis plan that generated this page is on <a href="https://github.com/hakyimlab/QGT-Columbia-HKI/blob/master/analysis/analysis_plan.Rmd">github here</a></p>
</div>
<div id="transcriptome-wide-association-methods" class="section level1">
<h1>Transcriptome-wide association methods</h1>
<pre class="r"><code>print(getwd())</code></pre>
<pre><code>[1] &quot;/Users/yanyul/Documents/repo/github/QGT-Columbia-HKI&quot;</code></pre>
<pre class="r"><code>pre=&quot;/home/student/QGT-Columbia-HKI&quot;
model.dir=glue::glue(&quot;{pre}/models&quot;)
metaxcan.dir=glue::glue(&quot;{pre}/repos/MetaXcan-master/software&quot;)
fastenloc.dir=glue::glue(&quot;{pre}/repos/fastenloc-master&quot;)
torus.dir=glue::glue(&quot;{pre}/repos/torus-master&quot;)
twmr.dir=glue::glue(&quot;{pre}/repos/TWMR-master&quot;)
results.dir=glue::glue(&quot;{pre}/results&quot;)</code></pre>
<div id="predict-expression" class="section level2">
<h2>predict expression</h2>
<div class="figure">
<img src="../extras/figures/PrediXcan-run.png" alt="Visual summary of predixcan runs" />
<p class="caption">Visual summary of predixcan runs</p>
</div>
<pre class="bash"><code>
conda activate imlabtools

export PRE=&quot;/home/student/QGT-Columbia-HKI&quot;
export DATA=$PRE/data/predixcan
export MODEL=$PRE/models
export METAXCAN=$PRE/repos/MetaXcan-master/software
export RESULTS=$PRE/results

printf &quot;Predict expression\n\n&quot;
python3 $METAXCAN/Predict.py \
--model_db_path $PRE/models/gtex_v8_en/en_Whole_Blood.db \
--vcf_genotypes $DATA/genotype/ALL.chr22.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz \
--vcf_mode genotyped \
--variant_mapping $DATA/gtex_v8_eur_filtered_maf0.01_monoallelic_variants.txt.gz id rsid \
--on_the_fly_mapping METADATA &quot;chr{}_{}_{}_{}_b38&quot; \
--prediction_output $RESULTS/predixcan/Whole_Blood__predict.txt \
--prediction_summary_output $RESULTS/predixcan/Whole_Blood__summary.txt \
--verbosity 9 \
--throw</code></pre>
</div>
<div id="assess-prediction-performance-optional" class="section level2">
<h2>assess prediction performance (optional)</h2>
<pre class="r"><code>predicted_expression = read_tsv(glue::glue(&quot;{results.dir}/predixcan/Whole_Blood__predict.txt&quot;))
dim(predicted_expression)
head(predicted_expression[,1:5])
prediction_summary = read_tsv(glue::glue(&quot;{results.dir}/predixcan/Whole_Blood__summary.txt&quot;))
dim(prediction_summary)
head(prediction_summary)

## merge with GEUVADIS expression data

## calculate spearman correlation

## select a few genes and plot predicted vs observed expression</code></pre>
</div>
<div id="run-association-with-phenotype" class="section level2">
<h2>run association with phenotype</h2>
<pre class="bash"><code>
export PRE=&quot;/home/student/QGT-Columbia-HKI&quot;
export DATA=$PRE/data/predixcan
export MODEL=$PRE/models
export METAXCAN=$PRE/repos/MetaXcan-master/software
export RESULTS=$PRE/results

printf &quot;association\n\n&quot;
python3 $METAXCAN/PrediXcanAssociation.py \
--expression_file $RESULTS/predixcan/Whole_Blood__predict.txt \
--input_phenos_file $DATA/phenotype/random_pheno_1000G_hg38.txt \
--input_phenos_column pheno \
--output $RESULTS/predixcan/random_pheno/Whole_Blood__association.txt \
--verbosity 9 \
--throw
</code></pre>
</div>
<div id="read-results" class="section level2">
<h2>read results</h2>
<pre class="r"><code>predixcan_association = read_tsv(glue::glue(&quot;{results.dir}/predixcan/random_pheno/Whole_Blood__association.txt&quot;))
dim(predixcan_association)
predixcan_association %&gt;% arrange(pvalue) %&gt;% head
predixcan_association %&gt;% arrange(pvalue) %&gt;% ggplot(aes(pvalue)) + geom_histogram(bins=20)</code></pre>
</div>
<div id="exercise" class="section level2">
<h2>Exercise</h2>
<p>-[ ] Run association with another phenotype in $PRE/predixcan/data/phenotype/ALL.chr22.shapeit2_integrated_snvindels_v2a_27022019.GRCh38_x_en_Whole_Blood.simulated_phenotype.spike_n_slab_0.5_x_pve0.6.txt</p>
<pre class="bash"><code></code></pre>
</div>
<div id="section" class="section level2">
<h2>——-</h2>
</div>
</div>
<div id="summary-predixcan" class="section level1">
<h1>Summary PrediXcan</h1>
<div class="figure">
<img src="../extras/figures/gwas-PrediXcan-spredixcan.png" alt="Visual summary of s-predixcan" />
<p class="caption">Visual summary of s-predixcan</p>
</div>
<pre class="r"><code>## harmonized and imputed GWAS result for coronary artery disease is available in 
# $PRE/s-predixcan/data/</code></pre>
<div id="run-s-predixcan" class="section level2">
<h2>run s-predixcan</h2>
<pre class="bash"><code>
export PRE=&quot;/home/student/QGT-Columbia-HKI&quot;
export DATA=$PRE/data/s-predixcan
export MODEL=$PRE/models
export METAXCAN=$PRE/repos/MetaXcan-master/software
export RESULTS=$PRE/results

python $METAXCAN/SPrediXcan.py \
--gwas_file  $DATA/imputed_CARDIoGRAM_C4D_CAD_ADDITIVE.txt.gz \
--snp_column panel_variant_id --effect_allele_column effect_allele --non_effect_allele_column non_effect_allele --zscore_column zscore \
--model_db_path $MODEL/gtex_v8_mashr/mashr_Whole_Blood.db \
--covariance $MODEL/gtex_v8_mashr/mashr_Whole_Blood.txt.gz \
--keep_non_rsid --additional_output --model_db_snp_key varID \
--throw \
--output_file $RESULTS/spredixcan/eqtl/CARDIoGRAM_C4D_CAD_ADDITIVE__PM__Whole_Blood.csv
</code></pre>
</div>
<div id="plot-and-interpret-s-predixcan-results" class="section level2">
<h2>plot and interpret s-predixcan results</h2>
<pre class="r"><code>spredixcan_association = read_csv(glue::glue(&quot;{results.dir}/spredixcan/eqtl/CARDIoGRAM_C4D_CAD_ADDITIVE__PM__Whole_Blood.csv&quot;))
dim(spredixcan_association)
spredixcan_association %&gt;% arrange(pvalue) %&gt;% head
spredixcan_association %&gt;% arrange(pvalue) %&gt;% ggplot(aes(pvalue)) + geom_histogram(bins=20)</code></pre>
<p>SORT1, considered to be a causal gene for LDL cholesterol and as a consequence of coronary artery disease, is not found here. Why? (tissue)</p>
</div>
<div id="run-multixcan-optional" class="section level2">
<h2>run multixcan (optional)</h2>
<pre class="bash"><code>
export MODEL=$PRE/models
export DATA=$PRE/data/s-predixcan

python $METAXCAN/SMulTiXcan.py \
--models_folder $MODEL/gtex_v8_mashr \
--models_name_pattern &quot;mashr_(.*).db&quot; \
--snp_covariance $MODEL/gtex_v8_expression_mashr_snp_covariance.txt.gz \
--metaxcan_folder $RESULTS/spredixcan/eqtl/ \
--metaxcan_filter &quot;CARDIoGRAM_C4D_CAD_ADDITIVE__PM__(.*).csv&quot; \
--metaxcan_file_name_parse_pattern &quot;(.*)__PM__(.*).csv&quot; \
--gwas_file $DATA/imputed_CARDIoGRAM_C4D_CAD_ADDITIVE.txt.gz \
--snp_column panel_variant_id --effect_allele_column effect_allele --non_effect_allele_column non_effect_allele --zscore_column zscore --keep_non_rsid --model_db_snp_key varID \
--cutoff_condition_number 30 \
--verbosity 7 \
--throw \
--output $RESULTS/smultixcan/eqtl/CARDIoGRAM_C4D_CAD_ADDITIVE_smultixcan.txt
</code></pre>
</div>
</div>
<div id="colocalization-methods" class="section level1">
<h1>Colocalization methods</h1>
<div class="figure">
<img src="../extras/figures/colocalization-run.png" alt="Visual summary of colocalization" />
<p class="caption">Visual summary of colocalization</p>
</div>
<div id="fine-map-gwas-results" class="section level2">
<h2>fine-map GWAS results</h2>
<p>We will run torus due to time limitation but ideally we would like to run a method that allows multiple causal variants per locus.</p>
<pre class="bash"><code>
#torus -d Height.torus.zval.gz --load_zval -dump_pip Height.gwas.pip
#gzip Height.gwas.pip
TORUSOFT=torus

$TORUSOFT -d $PRE/data/fastenloc/Height.torus.zval.gz --load_zval -dump_pip $PRE/data/fastenloc/Height.gwas.pip
cd $PRE/data/fastenloc
gzip Height.gwas.pip
cd $PRE 
</code></pre>
</div>
<div id="calculate-colocalization-with-fastenloc" class="section level2">
<h2>calculate colocalization with fastENLOC</h2>
<pre class="bash"><code>## check out tutorial https://github.com/xqwen/fastenloc/tree/master/tutorial

export eqtl_annotation_gzipped=$PRE/data/fastenloc/FASTENLOC-gtex_v8.eqtl_annot.vcf.gz
export gwas_data_gzipped=$PRE/data/fastenloc/Height.gwas.pip.gz
export TISSUE=Whole_Blood
export FASTENLOCSOFT=fastenloc

mkdir $RESULTS/fastenloc/
cd $RESULTS/fastenloc/
$FASTENLOCSOFT -eqtl $eqtl_annotation_gzipped -gwas $gwas_data_gzipped -t $TISSUE 

#[-total_variants total_snp] [-thread n] [-prefix prefix_name] [-s shrinkage]
</code></pre>
</div>
<div id="analyze-results" class="section level2">
<h2>analyze results</h2>
<pre class="r"><code>## optional - compare with s-predixcan results</code></pre>
<p>-[] prepare</p>
<hr />
</div>
</div>
<div id="mendelian-randomization-methods" class="section level1">
<h1>Mendelian randomization methods</h1>
<div id="run-smr-optional" class="section level2">
<h2>run SMR (optional)</h2>
<pre class="bash"><code></code></pre>
</div>
<div id="run-twmr-for-a-locus" class="section level2">
<h2>run TWMR (for a locus)</h2>
<div class="figure">
<img src="../extras/figures/TWMR.png" alt="TWMR" />
<p class="caption">TWMR</p>
</div>
<pre class="bash"><code>
export TWMR=$PRE/repos/TWMR-master
export OUTPUT=$PRE/results
GENE=ENSG00000002919

cd $TWMR

R &lt; $TWMR/MR.R --no-save $GENE

cd $PRE

## output: /home/student/QGT-Columbia-HKI/repos/TWMR-master/ENSG00000002919.alpha
</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] forcats_0.4.0   stringr_1.4.0   dplyr_0.8.3     purrr_0.3.3    
 [5] readr_1.3.1     tidyr_1.0.2     tibble_2.1.3    ggplot2_3.2.1  
 [9] tidyverse_1.3.0 workflowr_1.6.0

loaded via a namespace (and not attached):
 [1] tidyselect_1.0.0 xfun_0.12        haven_2.2.0      lattice_0.20-38 
 [5] colorspace_1.4-1 vctrs_0.2.2      generics_0.0.2   htmltools_0.4.0 
 [9] yaml_2.2.0       rlang_0.4.3      later_1.0.0      pillar_1.4.3    
[13] withr_2.1.2      glue_1.3.1       DBI_1.1.0        dbplyr_1.4.2    
[17] modelr_0.1.5     readxl_1.3.1     lifecycle_0.1.0  munsell_0.5.0   
[21] gtable_0.3.0     cellranger_1.1.0 rvest_0.3.5      evaluate_0.14   
[25] knitr_1.27       httpuv_1.5.2     fansi_0.4.1      broom_0.5.4     
[29] Rcpp_1.0.3       promises_1.1.0   backports_1.1.5  scales_1.1.0    
[33] jsonlite_1.6     fs_1.3.1         hms_0.5.3        digest_0.6.23   
[37] stringi_1.4.5    grid_3.6.2       rprojroot_1.3-2  cli_2.0.1       
[41] tools_3.6.2      magrittr_1.5     lazyeval_0.2.2   crayon_1.3.4    
[45] whisker_0.4      pkgconfig_2.0.3  xml2_1.2.2       reprex_0.3.0    
[49] lubridate_1.7.4  rstudioapi_0.10  assertthat_0.2.1 rmarkdown_2.1   
[53] httr_1.4.1       R6_2.4.1         nlme_3.1-143     git2r_0.26.1    
[57] compiler_3.6.2  </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
