---
title: "Evaluation of performance of PTRS/OTRS by breast cancer subtype"
author: "Padma Sheila Rajagopal, MD MPH"
date: "1/20/2020"

site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    df_print: paged
    toc: true
    theme: spacelab
    highlight: textmate
---
```{r DATA setup, warning=FALSE, include=FALSE, echo=FALSE}
rm(list = ls())
library(tidyverse)
library(dplyr)
library(ggplot2)
library(data.table)
library(pander)
library(ggExtra)
library(survival)
library(survminer)
library(stringr)
library(glmnet)
library(survival)
library(factoextra)
library(flashClust)
knitr::opts_chunk$set(autodep = TRUE)
options(datatable.fread.datatable = F)
options(stringsAsFactors = F)
panderOptions('table.split.table', Inf)
predixcan_cutoffs = c(5e-8, 1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1)

setwd("~/Research-Local/2020-ptrsotrs")
load("Input/PTRSOTRS.RData")

gen.score <- function(df, x, a, b) { 
  ## extract the objective column
  df1 = df[,a:b]
  col = df[, x]
  df1 * col
}

calculate_trs = function(expression_df, effect_size_df, pvalue_cutoffs = c(5e-8, 1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1)) {
  # expression_df should have rownames as gene id and colnames as individual id
  # effect_size_df should have 
  #     1. column named gene with gene id;
  #     2. column named effect_size as effect size;
  #     3. column named pvalue as pvalue
  
  # first of all, we remove genes with crazily large effect size (they will skew the PTRS distribution)
  effect_size_df_cleaned = effect_size_df %>% filter(abs(effect_size) < 5)
  
  # limit analysis to genes occcur in both expression_df and effect_size_df
  pred_expr_gene_pool = rownames(expression_df)
  predixcan_gene_pool = effect_size_df_cleaned$gene
  gene_pool_intersect = intersect(pred_expr_gene_pool, predixcan_gene_pool)
  
  # loop over p-value cutoffs and calculate PTRS at each cutoff
  out = list()
  outgenes = list()
  for(c in pvalue_cutoffs) {
    predixcan_i = effect_size_df_cleaned %>% filter(gene %in% gene_pool_intersect)
    beta_gene = predixcan_i %>% filter(pvalue < c) %>% select(gene, effect_size)
    outgenes[[length(outgenes) + 1]] = beta_gene %>% mutate(cutoff = c)
    indiv = colnames(expression_df) 
    mat = t(expression_df)
    mat = mat[, colnames(mat) %in% beta_gene$gene, drop = F]
    mat = mat[, match(beta_gene$gene, colnames(mat)), drop = F]
    tprs_i = as.matrix(mat) %*% beta_gene$effect_size
    df_tprs = data.frame(indiv = indiv, tprs = tprs_i, cutoff = c)
    rownames(df_tprs) = NULL
    out[[length(out) + 1]] = df_tprs
  }
  out = do.call(rbind, out)
  outgenes = do.call(rbind, outgenes)
  list(ptrs = out, selected_genes = outgenes)
}
```

## LumA/ Predicted (PTRS)
```{r LumA-specific Predicted-TPRS score generation for overall survival, cache=TRUE, echo=FALSE, warning=FALSE}
setwd("~/Research-Local/2020-ptrsotrs")

ospredictedexpressionLumA <- read.table("Input/predicted_expression/BRCA/os-results_predicted_expression.txt", header=TRUE, as.is = TRUE)
 
LumA <- subtype_formatted %>% filter(pam50=="Luminal A")
LumA <- LumA$submitter_id
HRpos <- subtype_formatted %>% filter(subtypefinal=="HR+", is.na(pam50)==TRUE)
HRpos <- HRpos$submitter_id

PTPRSLumAweights <- read.table("Input/risk_weights/2020_BRCA_risk_weights/BCAC2020_LumA_spredixcan_GTExV8_mashr.csv", header=TRUE, sep=",")
PTPRSLumAweights <- PTPRSLumAweights %>% dplyr::select(gene, gene_name, effect_size, pvalue)

ospredictedexpressionLumA <- ospredictedexpressionLumA %>% filter((IID %in% LumA)==TRUE)
ospredictedexpressionLumA <- as.data.frame(t(ospredictedexpressionLumA))
names(ospredictedexpressionLumA) <- ospredictedexpressionLumA["IID",]
ospredictedexpressionLumA <- ospredictedexpressionLumA[-1,] #Removing the FID row
ospredictedexpressionLumA <- ospredictedexpressionLumA[-1,] #Removing the IID row 
ospredictedexpressionLumAgene <- rownames(ospredictedexpressionLumA)
ospredictedexpressionLumA <- as.data.frame(lapply(ospredictedexpressionLumA, as.numeric))
rownames(ospredictedexpressionLumA) <- ospredictedexpressionLumAgene
ospredictedexpressionLumA$gene <- row.names(ospredictedexpressionLumA)

ospredictedexpressionLumA <- merge(ospredictedexpressionLumA, PTPRSLumAweights, by="gene", all.x=TRUE)
PTRSLumAscoreraw <- gen.score(ospredictedexpressionLumA, 185, 2, 183)
PTRSLumAscoredev <- PTRSLumAscoreraw
rownames(PTRSLumAscoredev) <- ospredictedexpressionLumAgene

PTRSLumAoptimization <- calculate_trs(PTRSLumAscoredev, PTPRSLumAweights)

pheno <- famOS %>% select(fullID,phenotype)
names(pheno) = c("SUBJID", "phenotype") 

perf_t = list()
ptrs <- PTRSLumAoptimization$ptrs
ptrs$indiv = stringr::str_replace_all(ptrs$indiv, '\\.', '-')
for(c in predixcan_cutoffs) {
  perf_t[[as.character(c)]] = inner_join(pheno, ptrs %>% dplyr::filter(cutoff == c), by = c('SUBJID' = 'indiv'))
}

for(i in names(perf_t)) {
    p = perf_t[[i]] %>% ggplot() + geom_density(aes(x = tprs, fill = factor(phenotype)), alpha = .5) + theme(legend.position = 'bottom')
    p = p + ggtitle(paste0('p-value cutoff: ', i))
    ggsave(paste0('~/Desktop/', 'ptrs_cutoff_', i, '.pdf'), p, height = 10, width = 6)
  }

PTRSLumAscorefinal <- PTRSLumAoptimization$ptrs %>% dplyr::filter(cutoff==1e-4)
PTRSLumAscorefinal$indiv = stringr::str_replace_all(PTRSLumAscorefinal$indiv, '\\.', '-')
PTRSLumAscorefinal <- merge(PTRSLumAscorefinal, famOS, by.x="indiv", by.y="fullID")
PTRSLumAscorefinal <- PTRSLumAscorefinal %>% select (indiv, tprs, cutoff, phenotype)
```

```{r LumA-specific Cox regression of Predicted TRS and overall survival, cache=TRUE, echo=FALSE, warning=FALSE}
survivalwhitePTRSLumA <- merge(PTRSLumAscorefinal, TCGAavail, by.x="indiv", by.y="submitter_id")
survivalwhitePTRSLumA$phenotype <- ifelse(survivalwhitePTRSLumA$phenotype=="2", "case", survivalwhitePTRSLumA$phenotype)
survivalwhitePTRSLumA$phenotype <- ifelse(survivalwhitePTRSLumA$phenotype=="1", 0, survivalwhitePTRSLumA$phenotype)
survivalwhitePTRSLumA$phenotype <- ifelse(survivalwhitePTRSLumA$phenotype=="case", 1, survivalwhitePTRSLumA$phenotype)

#Cox regression of overall survival
surv_objectPTRSLumA <- Surv(time=survivalwhitePTRSLumA$oss.months, event=as.numeric(survivalwhitePTRSLumA$phenotype))

fit.coxphPTRSLumA <- coxph(surv_objectPTRSLumA ~ survivalwhitePTRSLumA$tprs,
                   data = survivalwhitePTRSLumA)

ggforest(fit.coxphPTRSLumA, data = survivalwhitePTRSLumA, main="Predicted transcriptomic score with Luminal A GWAS + subtype in TCGA\n overall survival, EUR ancestry")
```

In a univariate analysis, there is no association between PTRS and OS when weighted by a Luminal-A specific GWAS in Luminal A patients. 

## LumB/ Predicted (PTRS)
```{r LumB-specific Predicted-TPRS score generation for overall survival, echo=FALSE, cache=TRUE,  warning=FALSE}
setwd("~/Research-Local/2020-ptrsotrs")

ospredictedexpressionLumB <- read.table("Input/predicted_expression/BRCA/os-results_predicted_expression.txt", header=TRUE, as.is = TRUE)
LumB <- subtype_formatted %>% filter(pam50=="Luminal B")
LumB <- LumB$submitter_id
HRpos <- subtype_formatted %>% filter(subtypefinal=="HR+", is.na(pam50)==TRUE)
HRpos <- HRpos$submitter_id

PTPRSLumBweights <- read.table("Input/risk_weights/2020_BRCA_risk_weights/BCAC2020_LumB_spredixcan_GTExV8_mashr.csv", header=TRUE, sep=",")
PTPRSLumBweights <- PTPRSLumBweights %>% dplyr::select(gene, gene_name, effect_size, pvalue)

ospredictedexpressionLumB <- ospredictedexpressionLumB %>% filter((IID %in% LumB)==TRUE)
ospredictedexpressionLumB <- as.data.frame(t(ospredictedexpressionLumB))
names(ospredictedexpressionLumB) <- ospredictedexpressionLumB["IID",]
ospredictedexpressionLumB <- ospredictedexpressionLumB[-1,] #Removing the FID row
ospredictedexpressionLumB <- ospredictedexpressionLumB[-1,] #Removing the IID row 
ospredictedexpressionLumBgene <- rownames(ospredictedexpressionLumB)
ospredictedexpressionLumB <- as.data.frame(lapply(ospredictedexpressionLumB, as.numeric))
rownames(ospredictedexpressionLumB) <- ospredictedexpressionLumBgene
ospredictedexpressionLumB$gene <- row.names(ospredictedexpressionLumB)

ospredictedexpressionLumB <- merge(ospredictedexpressionLumB, PTPRSLumBweights, by="gene", all.x=TRUE)
PTRSLumBscoreraw <- gen.score(ospredictedexpressionLumB, 87, 2, 85)
PTRSLumBscoredev <- PTRSLumBscoreraw
rownames(PTRSLumBscoredev) <- ospredictedexpressionLumBgene

PTRSLumBoptimization <- calculate_trs(PTRSLumBscoredev, PTPRSLumBweights)

pheno <- famOS %>% select(fullID,phenotype)
names(pheno) = c("SUBJID", "phenotype") 

perf_t = list()
ptrs <- PTRSLumBoptimization$ptrs
ptrs$indiv = stringr::str_replace_all(ptrs$indiv, '\\.', '-')
for(c in predixcan_cutoffs) {
  perf_t[[as.character(c)]] = inner_join(pheno, ptrs %>% dplyr::filter(cutoff == c), by = c('SUBJID' = 'indiv'))
}

for(i in names(perf_t)) {
    p = perf_t[[i]] %>% ggplot() + geom_density(aes(x = tprs, fill = factor(phenotype)), alpha = .5) + theme(legend.position = 'bottom')
    p = p + ggtitle(paste0('p-value cutoff: ', i))
    ggsave(paste0('~/Desktop/', 'ptrs_cutoff_', i, '.pdf'), p, height = 10, width = 6)
  }

PTRSLumBscorefinal <- PTRSLumBoptimization$ptrs %>% dplyr::filter(cutoff==1e-4)
PTRSLumBscorefinal$indiv = stringr::str_replace_all(PTRSLumBscorefinal$indiv, '\\.', '-')
PTRSLumBscorefinal <- merge(PTRSLumBscorefinal, famOS, by.x="indiv", by.y="fullID")
PTRSLumBscorefinal <- PTRSLumBscorefinal %>% select (indiv, tprs, cutoff, phenotype)
```

```{r LumB-specific Cox regression of Predicted TRS and overall survival, cache=TRUE, echo=FALSE, warning=FALSE}
survivalwhitePTRSLumB <- merge(PTRSLumBscorefinal, TCGAavail, by.x="indiv", by.y="submitter_id")
survivalwhitePTRSLumB$phenotype <- ifelse(survivalwhitePTRSLumB$phenotype=="2", "case", survivalwhitePTRSLumB$phenotype)
survivalwhitePTRSLumB$phenotype <- ifelse(survivalwhitePTRSLumB$phenotype=="1", 0, survivalwhitePTRSLumB$phenotype)
survivalwhitePTRSLumB$phenotype <- ifelse(survivalwhitePTRSLumB$phenotype=="case", 1, survivalwhitePTRSLumB$phenotype)

#Cox regression of overall survival
surv_objectPTRSLumB <- Surv(time=survivalwhitePTRSLumB$oss.months, event=as.numeric(survivalwhitePTRSLumB$phenotype))

fit.coxphPTRSLumB <- coxph(surv_objectPTRSLumB ~ survivalwhitePTRSLumB$tprs,
                   data = survivalwhitePTRSLumB)

ggforest(fit.coxphPTRSLumB, data = survivalwhitePTRSLumB, main="Predicted transcriptomic score with Luminal B GWAS + subtype in TCGA\n overall survival, EUR ancestry")
```

## TNBC/ Predicted (PTRS)
```{r TNBC-specific Predicted-TPRS score generation for overall survival, cache=TRUE, echo=FALSE, warning=FALSE}
setwd("~/Research-Local/2020-ptrsotrs")

ospredictedexpressionTNBC <- read.table("Input/predicted_expression/BRCA/os-results_predicted_expression.txt", header=TRUE, as.is = TRUE)
 
TNBC <- subtype_formatted %>% filter(pam50=="Basal-like")
TNBC <- TNBC$submitter_id

PTPRSTNBCweights <- read.table("Input/risk_weights/2020_BRCA_risk_weights/BCAC2020_TNBC_spredixcan_GTExV8_mashr.csv", header=TRUE, sep=",")
PTPRSTNBCweights <- PTPRSTNBCweights %>% dplyr::select(gene, gene_name, effect_size, pvalue)

ospredictedexpressionTNBC <- ospredictedexpressionTNBC %>% filter((IID %in% TNBC)==TRUE)
ospredictedexpressionTNBC <- as.data.frame(t(ospredictedexpressionTNBC))
names(ospredictedexpressionTNBC) <- ospredictedexpressionTNBC["IID",]
ospredictedexpressionTNBC <- ospredictedexpressionTNBC[-1,] #Removing the FID row
ospredictedexpressionTNBC <- ospredictedexpressionTNBC[-1,] #Removing the IID row 
ospredictedexpressionTNBCgene <- rownames(ospredictedexpressionTNBC)
ospredictedexpressionTNBC <- as.data.frame(lapply(ospredictedexpressionTNBC, as.numeric))
rownames(ospredictedexpressionTNBC) <- ospredictedexpressionTNBCgene
ospredictedexpressionTNBC$gene <- row.names(ospredictedexpressionTNBC)

ospredictedexpressionTNBC <- merge(ospredictedexpressionTNBC, PTPRSTNBCweights, by="gene", all.x=TRUE)
PTRSTNBCscoreraw <- gen.score(ospredictedexpressionTNBC, 76, 2, 74)
PTRSTNBCscoredev <- PTRSTNBCscoreraw
rownames(PTRSTNBCscoredev) <- ospredictedexpressionTNBCgene

PTRSTNBCoptimization <- calculate_trs(PTRSTNBCscoredev, PTPRSTNBCweights)

pheno <- famOS %>% select(fullID,phenotype)
names(pheno) = c("SUBJID", "phenotype") 

perf_t = list()
ptrs <- PTRSTNBCoptimization$ptrs
ptrs$indiv = stringr::str_replace_all(ptrs$indiv, '\\.', '-')
for(c in predixcan_cutoffs) {
  perf_t[[as.character(c)]] = inner_join(pheno, ptrs %>% dplyr::filter(cutoff == c), by = c('SUBJID' = 'indiv'))
}

for(i in names(perf_t)) {
    p = perf_t[[i]] %>% ggplot() + geom_density(aes(x = tprs, fill = factor(phenotype)), alpha = .5) + theme(legend.position = 'bottom')
    p = p + ggtitle(paste0('p-value cutoff: ', i))
    ggsave(paste0('~/Desktop/', 'ptrs_cutoff_', i, '.pdf'), p, height = 10, width = 6)
  }

PTRSTNBCscorefinal <- PTRSTNBCoptimization$ptrs %>% dplyr::filter(cutoff==1e-4)
PTRSTNBCscorefinal$indiv = stringr::str_replace_all(PTRSTNBCscorefinal$indiv, '\\.', '-')
PTRSTNBCscorefinal <- merge(PTRSTNBCscorefinal, famOS, by.x="indiv", by.y="fullID")
PTRSTNBCscorefinal <- PTRSTNBCscorefinal %>% select (indiv, tprs, cutoff, phenotype)
```

```{r TNBC-specific Cox regression of Predicted TRS and overall survival, cache=TRUE, echo=FALSE, warning=FALSE}
survivalwhitePTRSTNBC <- merge(PTRSTNBCscorefinal, TCGAavail, by.x="indiv", by.y="submitter_id")
survivalwhitePTRSTNBC$phenotype <- ifelse(survivalwhitePTRSTNBC$phenotype=="2", "case", survivalwhitePTRSTNBC$phenotype)
survivalwhitePTRSTNBC$phenotype <- ifelse(survivalwhitePTRSTNBC$phenotype=="1", 0, survivalwhitePTRSTNBC$phenotype)
survivalwhitePTRSTNBC$phenotype <- ifelse(survivalwhitePTRSTNBC$phenotype=="case", 1, survivalwhitePTRSTNBC$phenotype)

#Cox regression of overall survival
surv_objectPTRSTNBC <- Surv(time=survivalwhitePTRSTNBC$oss.months, event=as.numeric(survivalwhitePTRSTNBC$phenotype))

fit.coxphPTRSTNBC <- coxph(surv_objectPTRSTNBC ~ survivalwhitePTRSTNBC$tprs,
                   data = survivalwhitePTRSTNBC)

ggforest(fit.coxphPTRSTNBC, data = survivalwhitePTRSTNBC, main="Predicted transcriptomic score with TNBC GWAS + subtype in TCGA\n overall survival, EUR ancestry")
```

## HER2/ Predicted (PTRS)
```{r HER2-specific Predicted-TPRS score generation for overall survival, cache=TRUE, echo=FALSE, warning=FALSE}
setwd("~/Research-Local/2020-ptrsotrs")

ospredictedexpressionHER2 <-  read.table("Input/predicted_expression/BRCA/os-results_predicted_expression.txt", header=TRUE, as.is = TRUE)
 
HER2 <- subtype_formatted %>% filter(pam50=="HER2-enriched")
HER2 <- HER2$submitter_id

PTPRSHER2weights <- read.table("Input/risk_weights/2020_BRCA_risk_weights/BCAC2020_HER2_spredixcan_GTExV8_mashr.csv", header=TRUE, sep=",")
PTPRSHER2weights <- PTPRSHER2weights %>% dplyr::select(gene, gene_name, effect_size, pvalue)

ospredictedexpressionHER2 <- ospredictedexpressionHER2 %>% filter((IID %in% HER2)==TRUE)
ospredictedexpressionHER2 <- as.data.frame(t(ospredictedexpressionHER2))
names(ospredictedexpressionHER2) <- ospredictedexpressionHER2["IID",]
ospredictedexpressionHER2 <- ospredictedexpressionHER2[-1,] #Removing the FID row
ospredictedexpressionHER2 <- ospredictedexpressionHER2[-1,] #Removing the IID row 
ospredictedexpressionHER2gene <- rownames(ospredictedexpressionHER2)
ospredictedexpressionHER2 <- as.data.frame(lapply(ospredictedexpressionHER2, as.numeric))
rownames(ospredictedexpressionHER2) <- ospredictedexpressionHER2gene
ospredictedexpressionHER2$gene <- row.names(ospredictedexpressionHER2)

ospredictedexpressionHER2 <- merge(ospredictedexpressionHER2, PTPRSHER2weights, by="gene", all.x=TRUE)
PTRSHER2scoreraw <- gen.score(ospredictedexpressionHER2, 40, 2, 38)
PTRSHER2scoredev <- PTRSHER2scoreraw
rownames(PTRSHER2scoredev) <- ospredictedexpressionHER2gene

PTRSHER2optimization <- calculate_trs(PTRSHER2scoredev, PTPRSHER2weights)

pheno <- famOS %>% select(fullID,phenotype)
names(pheno) = c("SUBJID", "phenotype") 

perf_t = list()
ptrs <- PTRSHER2optimization$ptrs
ptrs$indiv = stringr::str_replace_all(ptrs$indiv, '\\.', '-')
for(c in predixcan_cutoffs) {
  perf_t[[as.character(c)]] = inner_join(pheno, ptrs %>% dplyr::filter(cutoff == c), by = c('SUBJID' = 'indiv'))
}

for(i in names(perf_t)) {
    p = perf_t[[i]] %>% ggplot() + geom_density(aes(x = tprs, fill = factor(phenotype)), alpha = .5) + theme(legend.position = 'bottom')
    p = p + ggtitle(paste0('p-value cutoff: ', i))
    ggsave(paste0('~/Desktop/', 'ptrs_cutoff_', i, '.pdf'), p, height = 10, width = 6)
  }

PTRSHER2scorefinal <- PTRSHER2optimization$ptrs %>% dplyr::filter(cutoff==1e-4)
PTRSHER2scorefinal$indiv = stringr::str_replace_all(PTRSHER2scorefinal$indiv, '\\.', '-')
PTRSHER2scorefinal <- merge(PTRSHER2scorefinal, famOS, by.x="indiv", by.y="fullID")
PTRSHER2scorefinal <- PTRSHER2scorefinal %>% select (indiv, tprs, cutoff, phenotype)
```

```{r HER2-specific Cox regression of Predicted TRS and overall survival, cache=TRUE, echo=FALSE, warning=FALSE}
survivalwhitePTRSHER2 <- merge(PTRSHER2scorefinal, TCGAavail, by.x="indiv", by.y="submitter_id")
survivalwhitePTRSHER2$phenotype <- ifelse(survivalwhitePTRSHER2$phenotype=="2", "case", survivalwhitePTRSHER2$phenotype)
survivalwhitePTRSHER2$phenotype <- ifelse(survivalwhitePTRSHER2$phenotype=="1", 0, survivalwhitePTRSHER2$phenotype)
survivalwhitePTRSHER2$phenotype <- ifelse(survivalwhitePTRSHER2$phenotype=="case", 1, survivalwhitePTRSHER2$phenotype)

#Cox regression of overall survival
surv_objectPTRSHER2 <- Surv(time=survivalwhitePTRSHER2$oss.months, event=as.numeric(survivalwhitePTRSHER2$phenotype))

fit.coxphPTRSHER2 <- coxph(surv_objectPTRSHER2 ~ survivalwhitePTRSHER2$tprs,
                   data = survivalwhitePTRSHER2)

ggforest(fit.coxphPTRSHER2, data = survivalwhitePTRSHER2, main="Predicted transcriptomic score with HER2+ GWAS + subtype in TCGA\n overall survival, EUR ancestry")
```

## Combined subtype-specific contribution (PTRS)
```{r Subtype-specific Cox regression of Predicted TRS and overall survival, echo=FALSE, cache=TRUE, warning=FALSE}
setwd("~/Research-Local/2020-ptrsotrs")
ospredictedexpression <- read.table("Input/predicted_expression/BRCA/os-results_predicted_expression.txt", header=TRUE, as.is = TRUE)
ospredictedexpressionreference <- ospredictedexpression
PTRSOSgroup <- ospredictedexpressionreference$IID

PTPRSweights <- read.table("Input/risk_weights/2020_BRCA_risk_weights/BCAC_Overall_2020_spredixcan_GTExV8_mashr.csv", header=TRUE, sep=",")
PTPRSweights <- PTPRSweights %>% dplyr::select(gene, gene_name, effect_size, pvalue)

ospredictedexpression <- as.data.frame(t(ospredictedexpression))
names(ospredictedexpression) <- ospredictedexpression["IID",]

ospredictedexpression <- ospredictedexpression[-1,] #Removing the FID row
ospredictedexpression <- ospredictedexpression[-1,] #Removing the IID row 
ospredictedexpressiongene <- rownames(ospredictedexpression)
ospredictedexpression <- as.data.frame(lapply(ospredictedexpression, as.numeric))
rownames(ospredictedexpression) <- ospredictedexpressiongene
ospredictedexpression$gene <- row.names(ospredictedexpression)

ospredictedexpression <- merge(ospredictedexpression, PTPRSweights, by="gene", all.x=TRUE)

PTRSscoreraw <- gen.score(ospredictedexpression, 697, 2, 695)
PTRSscoredev <- PTRSscoreraw
rownames(PTRSscoredev) <- ospredictedexpressiongene

PTRSoptimization <- calculate_trs(PTRSscoredev, PTPRSweights)
pheno <- famOS %>% select(fullID,phenotype)
names(pheno) = c("SUBJID", "phenotype") 

perf_t = list()
ptrs <- PTRSoptimization$ptrs
ptrs$indiv = stringr::str_replace_all(ptrs$indiv, '\\.', '-')
for(c in predixcan_cutoffs) {
  perf_t[[as.character(c)]] = inner_join(pheno, ptrs %>% dplyr::filter(cutoff == c), by = c('SUBJID' = 'indiv'))
}

for(i in names(perf_t)) {
    p = perf_t[[i]] %>% ggplot() + geom_density(aes(x = tprs, fill = factor(phenotype)), alpha = .5) + theme(legend.position = 'bottom')
    p = p + ggtitle(paste0('p-value cutoff: ', i))
    ggsave(paste0('~/Desktop/', 'ptrs_cutoff_', i, '.pdf'), p, height = 10, width = 6)
  }

PTRSscorefinal <- PTRSoptimization$ptrs %>% dplyr::filter(cutoff==1e-5)
PTRSscorefinal$indiv = stringr::str_replace_all(PTRSscorefinal$indiv, '\\.', '-')
PTRSscorefinal <- merge(PTRSscorefinal, famOS, by.x="indiv", by.y="fullID")
PTRSscorefinal <- PTRSscorefinal %>% select (indiv, tprs, cutoff, phenotype)
survivalwhitePTRS <- merge(PTRSscorefinal, TCGAavail, by.x="indiv", by.y="submitter_id")
survivalwhitesubtype <- survivalwhitePTRS

temp1 <- PTRSLumAscorefinal %>% select(indiv, tprs)
survivalwhitesubtype <- merge(survivalwhitesubtype, temp1, by.x="indiv", by.y="indiv", all.x=TRUE)
names(survivalwhitesubtype)[names(survivalwhitesubtype) == "tprs.y"] <- "tprs.luma"

temp2 <- PTRSLumBscorefinal %>% select(indiv, tprs)
survivalwhitesubtype <- merge(survivalwhitesubtype, temp2, by.x="indiv", by.y="indiv", all.x=TRUE)
names(survivalwhitesubtype)[names(survivalwhitesubtype) == "tprs"] <- "tprs.lumb"

temp3 <- PTRSTNBCscorefinal %>% select(indiv, tprs)
survivalwhitesubtype <- merge(survivalwhitesubtype, temp3, by.x="indiv", by.y="indiv", all.x=TRUE)
names(survivalwhitesubtype)[names(survivalwhitesubtype) == "tprs"] <- "tprs.basal"

temp4 <- PTRSHER2scorefinal %>% select(indiv, tprs)
survivalwhitesubtype <- merge(survivalwhitesubtype, temp4, by.x="indiv", by.y="indiv", all.x=TRUE)
names(survivalwhitesubtype)[names(survivalwhitesubtype) == "tprs"] <- "tprs.her2"

survivalwhitesubtype$lumayn <- ifelse (survivalwhitesubtype$pam50=="Luminal A", 1, 0)
survivalwhitesubtype$lumbyn <- ifelse (survivalwhitesubtype$pam50=="Luminal B", 1, 0)
survivalwhitesubtype$basalyn <- ifelse (survivalwhitesubtype$pam50=="Basal-like", 1, 0) 
survivalwhitesubtype$her2yn <- ifelse (survivalwhitesubtype$pam50=="HER2-enriched", 1, 0) 
survivalwhitesubtype$intluma <- survivalwhitesubtype$tprs.x*survivalwhitesubtype$lumayn
survivalwhitesubtype$intlumb <- survivalwhitesubtype$tprs.x*survivalwhitesubtype$lumbyn
survivalwhitesubtype$intbasal <- survivalwhitesubtype$tprs.x*survivalwhitesubtype$basalyn
survivalwhitesubtype$inther2 <- survivalwhitesubtype$tprs.x*survivalwhitesubtype$her2yn

#Cox regression of overall survival by subtype
surv_objectPTRSbysubtype <- Surv(time=survivalwhitesubtype$oss.months, event=as.numeric(survivalwhitesubtype$phenotype))
fit.coxphPTRSbysubtype <- coxph(surv_objectPTRSbysubtype ~
  survivalwhitesubtype$tprs.x+
  survivalwhitesubtype$lumayn+
  survivalwhitesubtype$lumbyn+
  survivalwhitesubtype$basalyn+
  survivalwhitesubtype$her2yn+
  survivalwhitesubtype$intluma+
  survivalwhitesubtype$intlumb+
  survivalwhitesubtype$intbasal+
  survivalwhitesubtype$inther2,
                   data = survivalwhitesubtype)
ggforest(fit.coxphPTRSbysubtype, data = survivalwhitesubtype, main="Predicted transcriptomic score from all patients\n subtype as covariate/interaction, overall survival, EUR ancestry")
```

No specific subtype (which are drived from existing RNA subtype methods) drove the score performance (although luminal B had the strongest effect). RNA-based subtype resulted in decreased efficacy of score performance. 