---
title: "Evaluation of performance of PTRS/OTRS using survival-GWAS to ascertain effect size (instead of risk)"
author: "Padma Sheila Rajagopal, MD MPH"
date: "1/20/2020"

site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    df_print: paged
    toc: true
    theme: spacelab
    highlight: textmate
---
```{r DATA setup, warning=FALSE, include=FALSE, echo=FALSE}
rm(list = ls())
library(tidyverse)
library(dplyr)
library(ggplot2)
library(data.table)
library(pander)
library(ggExtra)
library(survival)
library(survminer)
library(stringr)
library(glmnet)
library(survival)
library(factoextra)
library(flashClust)
knitr::opts_chunk$set(autodep = TRUE)
options(datatable.fread.datatable = F)
options(stringsAsFactors = F)
panderOptions('table.split.table', Inf)
predixcan_cutoffs = c(5e-8, 1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1)

setwd("~/Research-Local/2020-ptrsotrs")
load("Input/PTRSOTRS.RData")

gen.score <- function(df, x, a, b) { 
  ## extract the objective column
  df1 = df[,a:b]
  col = df[, x]
  df1 * col
}

calculate_trs = function(expression_df, effect_size_df, pvalue_cutoffs = c(5e-8, 1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1)) {
  # expression_df should have rownames as gene id and colnames as individual id
  # effect_size_df should have 
  #     1. column named gene with gene id;
  #     2. column named effect_size as effect size;
  #     3. column named pvalue as pvalue
  
  # first of all, we remove genes with crazily large effect size (they will skew the PTRS distribution)
  effect_size_df_cleaned = effect_size_df %>% filter(abs(effect_size) < 5)
  
  # limit analysis to genes occcur in both expression_df and effect_size_df
  pred_expr_gene_pool = rownames(expression_df)
  predixcan_gene_pool = effect_size_df_cleaned$gene
  gene_pool_intersect = intersect(pred_expr_gene_pool, predixcan_gene_pool)
  
  # loop over p-value cutoffs and calculate PTRS at each cutoff
  out = list()
  outgenes = list()
  for(c in pvalue_cutoffs) {
    predixcan_i = effect_size_df_cleaned %>% filter(gene %in% gene_pool_intersect)
    beta_gene = predixcan_i %>% filter(pvalue < c) %>% select(gene, effect_size)
    outgenes[[length(outgenes) + 1]] = beta_gene %>% mutate(cutoff = c)
    indiv = colnames(expression_df) 
    mat = t(expression_df)
    mat = mat[, colnames(mat) %in% beta_gene$gene, drop = F]
    mat = mat[, match(beta_gene$gene, colnames(mat)), drop = F]
    tprs_i = as.matrix(mat) %*% beta_gene$effect_size
    df_tprs = data.frame(indiv = indiv, tprs = tprs_i, cutoff = c)
    rownames(df_tprs) = NULL
    out[[length(out) + 1]] = df_tprs
  }
  out = do.call(rbind, out)
  outgenes = do.call(rbind, outgenes)
  list(ptrs = out, selected_genes = outgenes)
}
```

# PTRS-Overall survival in breast cancer (survival weights)
```{r Survival-weighted Predicted-TPRS score generation for overall survival, echo=FALSE, cache=TRUE,  warning=FALSE}
setwd("~/Research-Local/2020-ptrsotrs")
ospredictedexpressionssurv <- read.table("Input/predicted_expression/BRCA/os-results_predicted_expression.txt", header=TRUE, as.is = TRUE)

PTPRSsurvweights <- read.table("Input/risk_weights/BCAC2019_survival_spredixcan_GTExV8_mashr.csv", header=TRUE, sep=",")
PTPRSsurvweights <- PTPRSsurvweights %>% dplyr::select(gene, gene_name, effect_size, pvalue)

ospredictedexpressionssurv <- as.data.frame(t(ospredictedexpressionssurv))
names(ospredictedexpressionssurv) <- ospredictedexpressionssurv["IID",]

ospredictedexpressionssurv <- ospredictedexpressionssurv[-1,] #Removing the FID row
ospredictedexpressionssurv <- ospredictedexpressionssurv[-1,] #Removing the IID row 
ospredictedexpressionssurvgene <- rownames(ospredictedexpressionssurv)
ospredictedexpressionssurv <- as.data.frame(lapply(ospredictedexpressionssurv, as.numeric))
rownames(ospredictedexpressionssurv) <- ospredictedexpressionssurvgene
ospredictedexpressionssurv$gene <- row.names(ospredictedexpressionssurv)

ospredictedexpressionssurv <- merge(ospredictedexpressionssurv, PTPRSsurvweights, by="gene", all.x=TRUE)

PTRSsurvscoreraw <- gen.score(ospredictedexpressionssurv, 697, 2, 695)

PTRSsurvscoredev <- PTRSsurvscoreraw
rownames(PTRSsurvscoredev) <- ospredictedexpressionssurvgene

PTRSsurvoptimization <- calculate_trs(PTRSsurvscoredev, PTPRSsurvweights)

pheno <- famOS %>% select(fullID,phenotype)
names(pheno) = c("SUBJID", "phenotype") 

perf_t = list()
ptrs <- PTRSsurvoptimization$ptrs
ptrs$indiv = stringr::str_replace_all(ptrs$indiv, '\\.', '-')
for(c in predixcan_cutoffs) {
  perf_t[[as.character(c)]] = inner_join(pheno, ptrs %>% dplyr::filter(cutoff == c), by = c('SUBJID' = 'indiv'))
}

for(i in names(perf_t)) {
    p = perf_t[[i]] %>% ggplot() + geom_density(aes(x = tprs, fill = factor(phenotype)), alpha = .5) + theme(legend.position = 'bottom')
    p = p + ggtitle(paste0('p-value cutoff: ', i))
    ggsave(paste0('~/Desktop/', 'ptrs_cutoff_', i, '.pdf'), p, height = 10, width = 6)
  }

PTRSsurvscorefinal <- PTRSsurvoptimization$ptrs %>% dplyr::filter(cutoff==1e-2)
PTRSsurvscorefinal$indiv = stringr::str_replace_all(PTRSsurvscorefinal$indiv, '\\.', '-')
PTRSsurvscorefinal <- merge(PTRSsurvscorefinal, famOS, by.x="indiv", by.y="fullID")
PTRSsurvscorefinal <- PTRSsurvscorefinal %>% select (indiv, tprs, cutoff, phenotype)
```

```{r Survival-weighted Cox regression of Predicted TRS and overall survival, echo=FALSE, cache=TRUE, warning=FALSE}
survivalwhitePTRSsurv <- merge(PTRSsurvscorefinal, TCGAavail, by.x="indiv", by.y="submitter_id")
survivalwhitePTRSsurv$phenotype <- ifelse(survivalwhitePTRSsurv$phenotype=="2", "case", survivalwhitePTRSsurv$phenotype)
survivalwhitePTRSsurv$phenotype <- ifelse(survivalwhitePTRSsurv$phenotype=="1", 0, survivalwhitePTRSsurv$phenotype)
survivalwhitePTRSsurv$phenotype <- ifelse(survivalwhitePTRSsurv$phenotype=="case", 1, survivalwhitePTRSsurv$phenotype)

#Cox regression of overall survival
surv_objectPTRSsurv <- Surv(time=survivalwhitePTRSsurv$oss.months, event=as.numeric(survivalwhitePTRSsurv$phenotype))

fit.coxphPTRSsurv <- coxph(surv_objectPTRSsurv ~ survivalwhitePTRSsurv$tprs,
                   data = survivalwhitePTRSsurv)

ggforest(fit.coxphPTRSsurv, data = survivalwhitePTRSsurv, main="Predicted transcriptomic score with survival GWAS\n overall survival, EUR ancestry")
```

In a univariate analysis, there is no significant association between PTRS (when using weights from a GWAS of overall survival in breast cancer) and overall survival in breast cancer. 

# PTRS-Progression-free survival in breast cancer (survival weights)
```{r Survival-weighted Predicted-TPRS score generation for progression-free survival, echo=FALSE, cache=TRUE, warning=FALSE}

setwd("~/Research-Local/2020-ptrsotrs")
pfspredictedexpressionsurv <- read.table("Input/predicted_expression/BRCA/pfs-results_predicted_expression.txt", header=TRUE, as.is = TRUE)

pfspredictedexpressionsurv <- as.data.frame(t(pfspredictedexpressionsurv))
names(pfspredictedexpressionsurv) <- pfspredictedexpressionsurv["IID",]

pfspredictedexpressionsurv <- pfspredictedexpressionsurv[-1,] #Removing the FID row
pfspredictedexpressionsurv <- pfspredictedexpressionsurv[-1,] #Removing the IID row 
pfspredictedexpressiongenesurv <- rownames(pfspredictedexpressionsurv)
pfspredictedexpressionsurv <- as.data.frame(lapply(pfspredictedexpressionsurv, as.numeric))
rownames(pfspredictedexpressionsurv) <- pfspredictedexpressiongenesurv
pfspredictedexpressionsurv$gene <- row.names(pfspredictedexpressionsurv)

pfspredictedexpressionsurv <- merge(pfspredictedexpressionsurv, PTPRSsurvweights, by="gene", all.x=TRUE)

PTRSsurvscoreraw <- gen.score(pfspredictedexpressionsurv, 697, 2, 695)

PTRSsurvscoredev <- PTRSsurvscoreraw
rownames(PTRSsurvscoredev) <- pfspredictedexpressiongenesurv

PTRSsurvoptimization <- calculate_trs(PTRSsurvscoredev, PTPRSsurvweights)

pheno <- famPFS %>% select(fullID,phenotype)
names(pheno) = c("SUBJID", "phenotype") 

perf_t = list()
ptrs <- PTRSsurvoptimization$ptrs
ptrs$indiv = stringr::str_replace_all(ptrs$indiv, '\\.', '-')
for(c in predixcan_cutoffs) {
  perf_t[[as.character(c)]] = inner_join(pheno, ptrs %>% dplyr::filter(cutoff == c), by = c('SUBJID' = 'indiv'))
}

for(i in names(perf_t)) {
    p = perf_t[[i]] %>% ggplot() + geom_density(aes(x = tprs, fill = factor(phenotype)), alpha = .5) + theme(legend.position = 'bottom')
    p = p + ggtitle(paste0('p-value cutoff: ', i))
    ggsave(paste0('~/Desktop/', 'ptrs_cutoff_', i, '.pdf'), p, height = 10, width = 6)
  }

PTRSsurvscorefinal <- PTRSsurvoptimization$ptrs %>% dplyr::filter(cutoff==1e-1)
PTRSsurvscorefinal$indiv = stringr::str_replace_all(PTRSsurvscorefinal$indiv, '\\.', '-')
PTRSsurvscorefinal <- merge(PTRSsurvscorefinal, famPFS, by.x="indiv", by.y="fullID")
PTRSsurvscorefinal <- PTRSsurvscorefinal %>% select (indiv, tprs, cutoff, phenotype)
```

```{r Survival-weighted Cox regression of Predicted TRS and recurrence, cache=TRUE, echo=FALSE, warning=FALSE}
recurrencewhitePTRSsurv <- merge(PTRSsurvscorefinal, TCGAavail, by.x="indiv", by.y="submitter_id")
recurrencewhitePTRSsurv$phenotype <- ifelse(recurrencewhitePTRSsurv$phenotype=="2", "case", recurrencewhitePTRSsurv$phenotype)
recurrencewhitePTRSsurv$phenotype <- ifelse(recurrencewhitePTRSsurv$phenotype=="1", 0, recurrencewhitePTRSsurv$phenotype)
recurrencewhitePTRSsurv$phenotype <- ifelse(recurrencewhitePTRSsurv$phenotype=="case", 1, recurrencewhitePTRSsurv$phenotype)

#Cox regression of recurrence
surv_objectPTRSsurv <- Surv(time=recurrencewhitePTRSsurv$pfs.months, event=as.numeric(recurrencewhitePTRSsurv$phenotype))

fit.coxphPTRSsurv <- coxph(surv_objectPTRSsurv ~ recurrencewhitePTRSsurv$tprs,
                   data = recurrencewhitePTRSsurv)

ggforest(fit.coxphPTRSsurv, data = recurrencewhitePTRSsurv, main="Predicted transcriptomic score with survival GWAS\n progression-free survival, EUR ancestry")
```

In a univariate analysis, there is no significant association between PTRS (when using weights from a GWAS of overall survival in breast cancer) and progression-free survival in breast cancer. 

# OTRS-Overall survival in breast cancer (survival weights)
```{r Survival-weighted Observed-TPRS score generation, cache=TRUE, warning=FALSE, echo=FALSE}
setwd("~/Research-Local/2020-ptrsotrs")
osobservedexpression <- read.table("Input/observed_expression/BRCA-obexp/TCGA-BRCA_mRNA.csv", header=TRUE, sep=",")

samplesEURmatch <- famOS$fullID

osobservedexpression$submitter_id_long <- substr(osobservedexpression$aliquot_barcode, 1, 16)
osobservedexpression$submitter_id <- substr(osobservedexpression$aliquot_barcode, 1, 12)
osobservedexpression_cleaned = osobservedexpression %>% dplyr::filter(osobservedexpression$submitter_id %in% samplesEURmatch)
osobservedexpression_cleaned = osobservedexpression_cleaned %>% dplyr::filter(substr(osobservedexpression_cleaned$submitter_id_long, 14, 16)=="01A") #Isolates to primary tumor only, 682 samples remaining

osobservedexpression_cleaned$aliquot_barcode <- NULL

osobservedexpressiongenes <- colnames(osobservedexpression_cleaned)
rownames(osobservedexpression_cleaned) <- osobservedexpression_cleaned$submitter_id
osobservedexpressionnames <- rownames(osobservedexpression_cleaned)
osobservedexpressiongenes <- osobservedexpressiongenes[1:17321]

osobservedexpression_cleaned <- as.data.frame(t(osobservedexpression_cleaned))
osobservedexpression_cleaned <- osobservedexpression_cleaned[1:17321,] #removes text rows and only expression data observed
osobservedexpression_cleaned <- as.data.frame(lapply(osobservedexpression_cleaned, as.numeric))
osobservedexpressiongenes <- osobservedexpressiongenes[1:17321]
rownames(osobservedexpression_cleaned) <- osobservedexpressiongenes

OTPRSsurvweights <- read.table("Input/risk_weights/BCAC2019_survival_spredixcan_GTExV8_mashr.csv", header=TRUE, sep=",")
OTPRSsurvweights <- OTPRSsurvweights %>% dplyr::select(gene, gene_name, effect_size, pvalue)

osobservedexpression_cleaned$gene_name <- row.names(osobservedexpression_cleaned)
osobservedexpression_reference <- osobservedexpression_cleaned 

osobservedexpression_cleaned <- merge(osobservedexpression_cleaned, OTPRSsurvweights, by="gene_name", all.x=TRUE)
osobservedexpression_cleaned <- osobservedexpression_cleaned %>% dplyr::filter(!is.na(effect_size))
osobservedexpression_cleaned <- osobservedexpression_cleaned %>% dplyr::filter(duplicated(gene_name)==FALSE)

osobservedexpressiongenes <- osobservedexpression_cleaned$gene
osobservedexpressionrows <- osobservedexpression_cleaned$gene_name

osobservedexpression_cleaned_t <- osobservedexpression_cleaned %>%
  mutate_at(vars(-gene_name, -gene, -effect_size, -pvalue), function(x) {log(x+1)})

OTRSsurvscore <- gen.score(osobservedexpression_cleaned_t, 685, 2, 683)

OTRSsurvscoredev <- OTRSsurvscore
rownames(OTRSsurvscoredev) <- osobservedexpressiongenes

OTRSsurvoptimization <- calculate_trs(OTRSsurvscoredev, OTPRSsurvweights)

pheno <- famOS %>% select(fullID,phenotype)
names(pheno) = c("SUBJID", "phenotype") 

perf_o = list()
otrs <- OTRSsurvoptimization$ptrs
otrs$indiv = stringr::str_replace_all(otrs$indiv, '\\.', '-')
for(c in predixcan_cutoffs) {
  perf_o[[as.character(c)]] = inner_join(pheno, otrs %>% dplyr::filter(cutoff == c), by = c('SUBJID' = 'indiv'))
}

for(i in names(perf_o)) {
    p = perf_o[[i]] %>% ggplot() + geom_density(aes(x = tprs, fill = factor(phenotype)), alpha = .5) + theme(legend.position = 'bottom')
    p = p + ggtitle(paste0('p-value cutoff: ', i))
    ggsave(paste0('~/Desktop/', 'otrs_cutoff_', i, '.pdf'), p, height = 10, width = 6)
  }

OTRSsurvscorefinal <- OTRSsurvoptimization$ptrs %>% filter(cutoff==1e-3)
OTRSsurvscorefinal$indiv = stringr::str_replace_all(OTRSsurvscorefinal$indiv, '\\.', '-')
OTRSsurvscorefinal <- merge(OTRSsurvscorefinal, famOS, by.x="indiv", by.y="fullID")
OTRSsurvscorefinal <- OTRSsurvscorefinal %>% select (indiv, tprs, cutoff, phenotype)
```

```{r Survival-weighted Cox regression of Observed TRS and overall survival, echo=FALSE, cache=TRUE, warning=FALSE}
survivalwhiteOTRSsurv <- merge(OTRSsurvscorefinal, TCGAavail, by.x="indiv", by.y="submitter_id")

#Cox regression of overall survival
surv_objectOTRSsurv <- Surv(time=survivalwhiteOTRSsurv$oss.months, event=as.numeric(survivalwhiteOTRSsurv$phenotype))

fit.coxphOTRSsurv <- coxph(surv_objectOTRSsurv ~ survivalwhiteOTRSsurv$tprs,
                   data = survivalwhiteOTRSsurv)

ggforest(fit.coxphOTRSsurv, data = survivalwhiteOTRSsurv, main="Observed transcriptomic score with survival GWAS\n overall survival, EUR ancestry")
```

In a univariate analysis, there is no significant association between OTRS (when using weights from a GWAS of overall survival in breast cancer) and overall survival in breast cancer. 

# OTRS-Progression-free survival in breast cancer (survival weights)
```{r Observed-TPRS score generation for progression free survival, cache=TRUE, echo=FALSE, warning=FALSE}
pheno <- famPFS %>% select(fullID,phenotype)
names(pheno) = c("SUBJID", "phenotype") 

perf_o = list()
otrs <- OTRSsurvoptimization$ptrs
otrs$indiv = stringr::str_replace_all(otrs$indiv, '\\.', '-')
for(c in predixcan_cutoffs) {
  perf_o[[as.character(c)]] = inner_join(pheno, otrs %>% dplyr::filter(cutoff == c), by = c('SUBJID' = 'indiv'))
}

for(i in names(perf_o)) {
    p = perf_o[[i]] %>% ggplot() + geom_density(aes(x = tprs, fill = factor(phenotype)), alpha = .5) + theme(legend.position = 'bottom')
    p = p + ggtitle(paste0('p-value cutoff: ', i))
    ggsave(paste0('~/Desktop/', 'otrs_cutoff_', i, '.pdf'), p, height = 10, width = 6)
  }

OTRSsurvscorefinal <- OTRSsurvoptimization$ptrs %>% filter(cutoff==1e-3)
OTRSsurvscorefinal$indiv = stringr::str_replace_all(OTRSsurvscorefinal$indiv, '\\.', '-')
OTRSsurvscorefinal <- merge(OTRSsurvscorefinal, famOS, by.x="indiv", by.y="fullID")
OTRSsurvscorefinal <- OTRSsurvscorefinal %>% select (indiv, tprs, cutoff, phenotype)
```

```{r Survival-weighted Cox regression of Observed TRS and progression-free survival, echo=FALSE, cache=TRUE, warning=FALSE}
recurrencewhiteOTRSsurv <- merge(OTRSsurvscorefinal, TCGAavail, by.x="indiv", by.y="submitter_id")

#Cox regression of recurrence
surv_objectOTRSsurv <- Surv(time=recurrencewhiteOTRSsurv$pfs.months, event=as.numeric(recurrencewhiteOTRSsurv$phenotype))

fit.coxphOTRSsurv <- coxph(surv_objectOTRSsurv ~ recurrencewhiteOTRSsurv$tprs,
                   data = recurrencewhiteOTRSsurv)

ggforest(fit.coxphOTRSsurv, data = recurrencewhiteOTRSsurv, main="Observed transcriptomic score with survival GWAS\n progression-free survival, EUR ancestry")
```

In a univariate analysis, OTRS is not significantly associated with progression-free survival in breast cancer. 